File: src/lib/utils/dataFreshness.ts
export type DataFreshness = 'live' | 'stale'

interface DataFreshnessOptions {
  lastUpdated?: Date | string | number | null
  refreshInterval: number
  multiplier?: number
}

export function getDataFreshness({
  lastUpdated,
  refreshInterval,
  multiplier = 2,
}: DataFreshnessOptions): DataFreshness {
  if (!lastUpdated || refreshInterval <= 0) {
    return 'stale'
  }

  const lastUpdatedMs =
    lastUpdated instanceof Date ? lastUpdated.getTime() : new Date(lastUpdated).getTime()

  if (Number.isNaN(lastUpdatedMs)) {
    return 'stale'
  }

  const thresholdMs = refreshInterval * multiplier
  const ageMs = Date.now() - lastUpdatedMs

  return ageMs > thresholdMs ? 'stale' : 'live'
}

File: ios/SiouxlandOnline/Utilities/DataFreshness.swift
import Foundation

// MARK: - Data Freshness Calculation
// Mirrors the web app's getDataFreshness() logic

enum DataStatus: Sendable, Equatable {
    case live
    case stale
    case error(String)
    case loading

    var label: String {
        switch self {
        case .live: "Live"
        case .stale: "Stale"
        case .error: "Error"
        case .loading: "Loading"
        }
    }
}

enum DataFreshness {
    /// Determine if data is live or stale.
    /// Threshold: stale after `refreshInterval * 3`
    static func status(
        lastUpdated: Date?,
        refreshInterval: TimeInterval
    ) -> DataStatus {
        guard let lastUpdated else { return .loading }

        let elapsed = Date.now.timeIntervalSince(lastUpdated)
        let threshold = refreshInterval * 3

        return elapsed > threshold ? .stale : .live
    }
}
